aws/resource_aws_appsync_resolver.go

```diff
@@ -0,0 +1,182 @@
+package aws
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/appsync"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsAppsyncResolver() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsAppsyncResolverCreate,
+		Read:   resourceAwsAppsyncResolverRead,
+		Update: resourceAwsAppsyncResolverUpdate,
+		Delete: resourceAwsAppsyncResolverDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"api_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"field": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"data_source": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"request_template": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"response_template": {
+				Type:     schema.TypeString,
+				Required: true, // documentation bug, the api returns 400 if this is not specified.
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsAppsyncResolverCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	input := &appsync.CreateResolverInput{
+		ApiId:                   aws.String(d.Get("api_id").(string)),
+		DataSourceName:          aws.String(d.Get("data_source").(string)),
+		TypeName:                aws.String(d.Get("type").(string)),
+		FieldName:               aws.String(d.Get("field").(string)),
+		RequestMappingTemplate:  aws.String(d.Get("request_template").(string)),
+		ResponseMappingTemplate: aws.String(d.Get("response_template").(string)),
+	}
+
+	mutexKey := fmt.Sprintf("appsync-schema-%s", d.Get("api_id").(string))
+	awsMutexKV.Lock(mutexKey)
+	defer awsMutexKV.Unlock(mutexKey)
+
+	_, err := retryOnAwsCode(appsync.ErrCodeConcurrentModificationException, func() (interface{}, error) {
+		return conn.CreateResolver(input)
+	})
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId(d.Get("api_id").(string) + "-" + d.Get("type").(string) + "-" + d.Get("field").(string))
+
+	return resourceAwsAppsyncResolverRead(d, meta)
+}
+
+func resourceAwsAppsyncResolverRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	apiID, typeName, fieldName, err := decodeAppsyncResolverID(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	input := &appsync.GetResolverInput{
+		ApiId:     aws.String(apiID),
+		TypeName:  aws.String(typeName),
+		FieldName: aws.String(fieldName),
+	}
+
+	resp, err := conn.GetResolver(input)
+	if err != nil {
+		return err
```

Comment:
We should return error context here for operators and code maintainers:

```suggestion
	return fmt.Errorf("error getting AppSync Resolver (%s): %s", d.Id(), err)
```